tools/clang/unittests/HLSL/DxilContainerTest.cpp

```diff
@@ -1762,6 +1763,296 @@ TEST_F(DxilContainerTest, DxcUtils_CreateReflection) {
+      0x8d536ca1,
+      0x0cca,
+      0x4956,
+      {0xa8, 0x37, 0x78, 0x69, 0x63, 0x75, 0x55, 0x84}};
+
+  // Check ShaderReflection interface versions
+  {
+    CComPtr<IDxcOperationResult> pResult;
+    CComPtr<IDxcResult> pResultV2;
+    CComPtr<IDxcBlob> pReflectionPart;
+    CComPtr<IDxcBlobEncoding> pSource;
+    CreateBlobFromText(shaderSource, &pSource);
+    VERIFY_SUCCEEDED(pCompiler->Compile(pSource, L"hlsl.hlsl", L"main",
+                                        L"ds_6_0", nullptr, 0, nullptr, 0,
+                                        nullptr, &pResult));
+    VERIFY_SUCCEEDED(pResult->QueryInterface(&pResultV2));
+    VERIFY_SUCCEEDED(pResultV2->GetOutput(
+        DXC_OUT_REFLECTION, IID_PPV_ARGS(&pReflectionPart), nullptr));
+    DxcBuffer buffer = {pReflectionPart->GetBufferPointer(),
+                        pReflectionPart->GetBufferSize(), 0};
+
+    // The interface version supported for QI should only be the one used for
+    // creating the original reflection interface.
+
+    { // Verify with initial interface ID3D12ShaderReflection
+      CComPtr<ID3D12ShaderReflection> pShaderReflection;
+      VERIFY_SUCCEEDED(
+          pUtils->CreateReflection(&buffer, IID_PPV_ARGS(&pShaderReflection)));
+
+      // Verify QI for same interface and IUnknown succeeds:
+      CComPtr<ID3D12ShaderReflection> pShaderReflection2;
+      VERIFY_SUCCEEDED(
+          pShaderReflection->QueryInterface(IID_PPV_ARGS(&pShaderReflection2)));
+      CComPtr<IUnknown> pUnknown;
+      VERIFY_SUCCEEDED(
+          pShaderReflection->QueryInterface(IID_PPV_ARGS(&pUnknown)));
+
+      // Verify QI for wrong version of interface fails:
+      CComPtr<ID3D12ShaderReflection> pShaderReflection43;
+      VERIFY_FAILED(pShaderReflection->QueryInterface(
+          IID_ID3D11ShaderReflection_43, (void **)&pShaderReflection43));
+      CComPtr<ID3D12ShaderReflection> pShaderReflection47;
+      VERIFY_FAILED(pShaderReflection->QueryInterface(
+          IID_ID3D11ShaderReflection_47, (void **)&pShaderReflection47));
+
+      // Verify QI for wrong interface fails:
+      CComPtr<ID3D12LibraryReflection> pLibraryReflection;
+      VERIFY_FAILED(
+          pShaderReflection->QueryInterface(IID_PPV_ARGS(&pLibraryReflection)));
+
+      // Verify Get*ParameterDesc methods do not write out-of-bounds
+      TestDescStruct testParamDesc;
+      VERIFY_SUCCEEDED(pShaderReflection->GetInputParameterDesc(
+          0, &testParamDesc.paramDesc));
+      VERIFY_IS_TRUE(testParamDesc.CheckBytesAfterParamDesc());
+      VERIFY_SUCCEEDED(pShaderReflection->GetOutputParameterDesc(
+          0, &testParamDesc.paramDesc));
+      VERIFY_IS_TRUE(testParamDesc.CheckBytesAfterParamDesc());
+      VERIFY_SUCCEEDED(pShaderReflection->GetPatchConstantParameterDesc(
+          0, &testParamDesc.paramDesc));
+      VERIFY_IS_TRUE(testParamDesc.CheckBytesAfterParamDesc());
+    }
+
+    { // Verify with initial interface IID_ID3D11ShaderReflection_47
```

Comment:
Was this intended?  The code below defines a CComPtr for a ID3D12ShaderReflection interface.  Then calls CreateReflection( ), asking for IID_ID3D11ShaderReflection_47.  That seems odd to me.

Reply:
Yes, this is intentional.  I don't have any other version of the interface defined in any header I would want to include, and it shouldn't matter.  I only use a common part of the interface and verify the versioned behavior of those methods.  The difference between interface versions comes down to an additional signature element field in the structure pointer passed to the `Get*ParameterDesc` methods.